experimental/Schemes/Tjurina.jl

```diff
@@ -0,0 +1,770 @@
+This computation is based on the Milnor number respectively Tjurina number.
+# Examples
+```jldoctest
+julia> R,(x,y) = QQ["x", "y"];
+
+julia> f = x^5 + y^5 + x^2*y^2;
+
+julia> X = HypersurfaceGerm(AffineScheme(quo(R, ideal(R, f))[1]), [0, 0]);
+
+julia> determinacy_bound(X)
+11
+
+julia> determinacy_bound(X, :right)
+12
+```
+"""
+function determinacy_bound(X::HypersurfaceGerm, equivalence::Symbol = :contact)
+  f = defining_ideal(X)[1]
+  return determinacy_bound(f, equivalence)
+end
+
+
+
+
+
+@doc raw"""
+    sharper_determinacy_bound(f::MPolyLocRingElem)
+    sharper_determinacy_bound(f::MPolyLocRingElem, equivalence::Symbol = :contact)
+
+Compute some determinacy bound of 'f' w.r.t. ':right' or ':contact' equivalence.
+Return -1 if not finitely determined. 
+By default computes w.r.t. contact equivalence.
+At the cost of a higher computation time this function computes in general 
+some sharper determinacy bound than the function determinacy_bound. 
+In characteristic 0 the computed bound is the determinacy or the determinacy plus one.
+# Examples
+```jldoctest
+julia> R,(x,y) = QQ["x", "y"];
+
+julia> L,_  = localization(R, complement_of_point_ideal(R, [0, 0]));
+
+julia> f = L(x^3+y^6);
+
+julia> determinacy_bound(f, :right)
+11
+
+julia> sharper_determinacy_bound(f, :right)
+6
+```
+"""
+function sharper_determinacy_bound(f::MPolyLocRingElem, equivalence::Symbol = :contact)
+  equivalence == :right || equivalence == :contact || error("Equivalence typ must be ':right' or ':contact'.")
+  ord_f = order_as_series(f)
+  ## if the order of f is 1, then f is right and contact equivalent to its 1-jet (smooth case)
+  ord_f != 1 || return 1  
+  is_finitelydetermined(f::MPolyLocRingElem, equivalence) ||  return -1
+  R = base_ring(parent(f))
+  m = ideal(gens(R))
+  ## shift to 0 for computation w.r.t. local order  
+  shift, _ = base_ring_shifts(parent(f))
+  a = shift(numerator(f))  
+  b = shift(denominator(f)) 
+  if equivalence == :right
+    if ord_f == 0     
+      ## A unit has the same right-determinacy as the power series without the constant term.
+      ## Therefore remove constant term. 
+      L, _ = localization(R, complement_of_point_ideal(R, [coefficient_ring(R)(0) for i = 1:ngens(R)]))
+      f_shifted = L(a//b) 
+      f_new = f_shifted-(constant_coefficient(a)//constant_coefficient(b))
+      return sharper_determinacy_bound(f_new, equivalence)
+    end
+    J = ideal(R,[derivative(a, R[i])*b - a*derivative(b, R[i]) for i in 1:nvars(R)])
+    I = m^2*J
+  else  ## equivalence == :contact
+    ord_f != 0 || return 0
+    I = m*a + m^2*jacobian_ideal(a)
+  end
+  G = standard_basis(I, ordering = negdeglex(parent(a)))
+  h = Singular.highcorner(G.gens.S)
+  l = total_degree(R(h))  
+  ## m^(l+1) \subseteq I
+  ## char. 0:                                         
+  characteristic(R) != 0 || (k = l)
+  ## pos. char.: 2*(l-1) - order_as_series(f) + 2
+  characteristic(R) == 0 || (k = 2*l - order_as_series(f))
+  return k
```

Comment:
```suggestion
  return characteristic(R) == 0 ? l : 2*l - order_as_series(f)
```

Reply:
Is this a typo? it should read ...` ? l :` .... shouldn't it?